src/app/components/Persist/MigrateV0StateForm.tsx

```diff
@@ -0,0 +1,208 @@
+import { Box } from 'grommet/es6/components/Box'
+import { Button } from 'grommet/es6/components/Button'
+import { Form } from 'grommet/es6/components/Form'
+import { Paragraph } from 'grommet/es6/components/Paragraph'
+import { persistActions } from 'app/state/persist'
+import React, { useEffect, useState } from 'react'
+import { useDispatch } from 'react-redux'
+import { useTranslation } from 'react-i18next'
+import { PasswordField } from 'app/components/PasswordField'
+import { preventSavingInputsToUserData } from 'app/lib/preventSavingInputsToUserData'
+import { LoginModalLayout } from './LoginModalLayout'
+import {
+  MigratingV0State,
+  decryptWithPasswordV0,
+  readAndMigrateLanguageV0,
+  readStorageV0,
+} from '../../../utils/walletExtensionV0'
+import { NoTranslate } from '../NoTranslate'
+import { uintToBase64, hex2uint } from '../../lib/helpers'
+import { CheckBox } from 'grommet/es6/components/CheckBox'
+import { FormField } from 'grommet/es6/components/FormField'
+import { AddressFormatter } from '../AddressFormatter'
+import { themeActions } from '../../../styles/theme/slice'
+import { RevealOverlayButton } from '../RevealOverlayButton'
+
+export function MigrateV0StateForm() {
+  const { t, i18n } = useTranslation()
+  const dispatch = useDispatch()
+  const [password, setPassword] = useState('')
+  const [enteredWrongPassword, setWrongPassword] = useState(false)
+  const [migratingV0State, setMigratingV0State] = useState<undefined | MigratingV0State>(undefined)
+  const [hasSavedMnemonic, setHasSavedMnemonic] = useState(false)
+
+  useEffect(() => {
+    // Old extension used light theme
+    dispatch(themeActions.changeTheme('light'))
+  }, [dispatch])
+
+  useEffect(() => {
+    // Switch to zh_CN if it was used in old extension. We don't have a language selector in migration UI.
+    i18n.changeLanguage(readAndMigrateLanguageV0())
+  }, [i18n])
+
+  const onSubmitPassword = async () => {
+    const storageV0 = await readStorageV0()
+    if (!storageV0) throw new Error('No v0 storage to migrate')
+    const decryptedPromise = decryptWithPasswordV0(password, storageV0)
+    const enteredWrongPassword = await decryptedPromise.then(
+      () => false,
+      () => true,
+    )
+    setWrongPassword(enteredWrongPassword)
+    if (enteredWrongPassword) return
+
+    const decrypted = await decryptedPromise
+    setMigratingV0State(decrypted)
+  }
+
+  const onSubmitSavedMnemonic = () => {
+    if (!migratingV0State) throw new Error('No v0 storage to migrate')
+    if (migratingV0State.invalidPrivateKeys.length > 0) {
+      setHasSavedMnemonic(true)
+    } else {
+      onSubmitSavedInvalidPrivateKeys()
+    }
+  }
+
+  const onSubmitSavedInvalidPrivateKeys = () => {
+    finishMigration()
+  }
+
+  const finishMigration = () => {
+    if (!migratingV0State) throw new Error('No v0 storage to migrate')
+    dispatch(
+      persistActions.finishV0Migration({
+        persistedRootState: migratingV0State.state,
```

Comment:
It does shallow-merge persisted state with non-persisted state.
Non-persisted: just keeps initialized redux slices.
Persisted: overrides previous persisted redux slices.

I don't think there's much risk. It could potentially start persisting v1 and terminate before deleting v0. In this case it would behave like a successful migration, except that, if user deletes v1 profile, v0 migration flow will appear again.